### ISSUE TITLE
Add api config to customize API Gateway Responses

### ISSUE THREAD

**Issue by felipe-lee** [aws/chalice#1282]
There is a need and use for customizing API Gateway Responses in Chalice. Specifically, I mean the ones you can currently edit in the console on this page:

![image](https://user-images.githubusercontent.com/35938642/67136250-5a874800-f1e9-11e9-8835-804d328bdae3.png)

So you can set custom status response codes, headers, and possibly templates.

I'd started #1270 to try implementing it, but @stealthycoin it would be good to discuss the implementation in an issue. In the pr I tried implementing it in the chalice config, but as was pointed out, it would be more appropriate for this to be managed on the app itself, e.g.
```python
app = Chalice(app_name="helloworld")
app.api.<yet to be named> = <some config we haven't defined yet>
```

So some things we need to address in this issue are:
1. What should the name of the property on `app.api` be?
    - I personally like something like `gateway_responses`, but am open to other people's input.
2. How should this configuration be defined? There is the option of JSON, which as pointed out in the pr, is not as friendly. @stealthycoin suggested using a dict to define it. Here is a link to the possible values that we could define: https://docs.aws.amazon.com/apigateway/api-reference/resource/gateway-response/#properties
    - I agree that this should be a dict, and we could make it a mapping of response types to config options. E.g.
        ```python
        app.api.gateway_responses = {
            'default_4xx': {
                'response_parameters': {
                    'access-control-allow-origin': "'*'"
                }
            } 
        }
        ```
      Maybe this is too verbose? I'd be happy to get some feedback or alternative options.

**Comment by felipe-lee** [comment]
Forgot to say that one of the big reasons for being able to do this is so that you can set up Default 4XX and 5XX response parameters for CORS setup.

**Comment by 3dw1np** [comment]
Will be good to have this !

**Comment by Cullimoreorless** [comment]
Just found a different solution for this, if you use a Merge Template with cloud formation, as described here: [https://aws.github.io/chalice/topics/cfn.html#template-merge](url) then you can add properties to your generated RestAPI like so:

`{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Description": "this is the merge.json file used with the --merge-template command",
  "Resources": {
    "APIHandler": {
      "Type": "AWS::Serverless::Function"
    },
    "RestAPI": {
      "Properties": {
        "Cors": {
          "AllowHeaders": "'Selected-Org-Id,Is-Daylight-Savings,Pl-Timestamp,Primary-Org-Id,Sec-Fetch-Dest,Timezone-Offset'",
          "AllowOrigin": "'*'"
        },
        "GatewayResponses": {
          "DEFAULT_4XX": {
            "ResponseParameters": {
              "Headers": {
                "gatewayresponse.header.Access-Control-Allow-Origin": "'*'",
                "gatewayresponse.header.Access-Control-Allow-Headers": "'*'"
              }
            },
            "ResponseTemplates": {
              "application/json": "{\"message\": \"Default 4XX error\" }"
            }
          }
        }
      }
    },
......
}`

It's not part of the chalice code, I agree this would be great to have in the Chalice code itself, but this is a solution until then

**Comment by subssn21** [comment]
Not having this just bit me big time.

I deployed a change that I was working on and suddenly I was getting CORS errors and everything was working fine before and none of my changes should have affected CORS. It ends up after digging through API Gateway logs that I had a module dependency that didn't get loaded properly and was causing the app to die with:

errorMessage": "Unable to import module 'app': No module named 'pyheif'",
    "errorType": "Runtime.ImportModuleError"

API Gateway tried to return a 502 Gateway error, but the response I got on the browser was:

Access to XMLHttpRequest at 'https://xxxxxxxxxx.execute-api.us-east-2.amazonaws.com/dev/login' from origin 'https://localhost:8080' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

If anyone has ideas on what we should do, I would be happy to help work on this.

**Comment by apfromiit** [comment]
+1

I use chalice as backend and angular as frontend for my apps. Recently started using chalice authorizer. For unauthorized requests, Chalice authorizer always returns status 0 (CORS error) while expected to get Http 401 or 403 status code. As of now, I can fix the issue by enabling cors with DEFAULT_4xx Gateway Response in AWS API Gateway console and redeploying the apis.
 
 Though fix might exists via **[chalice package](https://github.com/aws/chalice/issues/1282#issuecomment-696761819)**, I would like to continue using **chalice deploy**. 
 
 Having this api config to customize API Gateway response is a must need for all Angular/React/web apps to receive 4xx (Unauthorized) errors when using chalice authorizer in chalice apps.
 
I tried achieving the same via awscli and boto3 without any success.

Like to know when will this feature be available in chalice.

**Comment by alexevansigg** [comment]
+1  for this requirement.

In my scenario i'm using a customer Lambda Authoriser... if the lambda authorizor rejects the users  then it will trigger the default API responses, and without a means to customize the Response template, there's no means to return meaningful response details.
